### PR TITLE
Switch from Gson to Jackson, and improve JSON responses

### PR DESCRIPTION
**Pull request by sudarsanm94**
PR #53 with some modifications.

### CODE REVIEW

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java [hunk 1]
```diff
@@ -203,17 +201,19 @@ public Response stop(StopType stopType)
...
     {
+        URI uri;
         try {
-            UriBuilder uriBuilder = fromUri(format("http://localhost:%s", informer.getPrestoPort())).path("/v1/info/state");
-            Response response = client.target(uriBuilder.build()).request(TEXT_PLAIN)
-                    .buildPut(entity(GSON.toJson("SHUTTING_DOWN"), APPLICATION_JSON)).invoke();
-            if (response.getStatus() != 200) {
-                throw new PrestoManagerException("Failed to stop presto gracefully");
-            }
+            uri = fromUri("http://localhost").port(informer.getPrestoPort())
```
Comment (maciejgrzybek) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: Nit but I'd put either all `.methods` in separate lines or all in one.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java [hunk 2]
```diff
@@ -203,17 +201,19 @@ public Response stop(StopType stopType)
...
+                    .path("/v1/info/state").build();
         }
         catch (IOException e) {
             throw new PrestoManagerException("Failed to get Presto port number.", e);
         }
+        Response response = client.target(uri).request(TEXT_PLAIN)
+                .buildPut(entity("\"SHUTTING_DOWN\"", APPLICATION_JSON)).invoke();
```
Comment (maciejgrzybek) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: Why these quote marks are necessary?

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Optional;
+
+/**
+ * Class for reading the entities of {@link Response Responses} into Java
+ * objects suitable to re-serialize in an outgoing {@link Response}
+ */
+public interface JsonResponseReader
```
Comment (maciejgrzybek) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java]: Why did you add this interface if there's only single implementation of it?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java [hunk 4]
```diff
@@ -234,60 +234,92 @@ public Response restart()
...
      */
     public Response status()
     {
-        try {
-            if (!isInstalled()) {
-                LOGGER.info("Presto is not installed");
-                return Response.status(OK).entity(ImmutableMap.of("installed", false)).type(APPLICATION_JSON)
+        ImmutableMap.Builder<String, Object> jsonBuilder = ImmutableMap.builder();
+        // Call this to get a return value on success
+        Supplier<Response> responseGetter = () ->
```
Comment (maciejgrzybek) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: What for is this supplier? Can't you call `jsonBuilder.build()` instead of `responseGetter.get()`?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java [hunk 5]
```diff
@@ -234,60 +234,92 @@ public Response restart()
...
-        // TODO: GET a JSON from that URI to get the real Presto version and other info
-        Response prestoInfo = client.target(uriBuilder.build()).request(APPLICATION_JSON).buildGet().invoke();
-        Response prestoState = client.target(uriBuilder.path("/state").build()).request(APPLICATION_JSON).buildGet().invoke();
 
-        ImmutableMap.Builder<String, Object> jsonBuilder = ImmutableMap.builder();
+        // TODO: Make the JSON response flat
+        Response prestoInfo = client.target(uriBuilder.build())
+                .request(APPLICATION_JSON).buildGet().invoke();
+        responseReader.readIfPossible(prestoInfo)
+                .ifPresent(o -> jsonBuilder.put("ServerInfo", o));
```
Comment (maciejgrzybek) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: Please don't use abbreviations

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java (hunk 2 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: This request to Presto should contain the body in `JSON` format.
For some reason Jersey client is not automatically serializing string to JSON format.
I should either have this quote marks or use `ObjectMapper` to serialize the string `SHUTTING_DOWN` to json format.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java (hunk 1 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: I'll change it

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java (hunk 3 above)
Comment (sudarsanm94) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java]: Could you answer this question @jirassimok?.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java (hunk 5 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: changed

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java (hunk 4 above)
Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: Both are similar.
`responseGetter.get()` is just shorter.

Comment (maciejgrzybek) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: But you have to define `Supplier` which is not necessary here. You introduce extra dependency which does not serve any purpose. It simply obfuscates the code.

Comment (sudarsanm94) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PackageController.java]: Changed

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java (hunk 3 above)
Comment (jirassimok) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/json/JsonResponseReader.java]: That was a mistake; I created the interface, then realized that it wasn't necessary, but I must have forgotten to undo the changes.
